nemo/collections/nlp/data/language_modeling/megatron/_megatron_batch_samplers.py

```diff
@@ -0,0 +1,193 @@
+import abc
+
+import torch
+
+__all__ = [
+    "MegatronPretrainingBatchSampler",
+    "MegatronPretrainingRandomBatchSampler",
+]
+
+
+class BaseMegatronBatchSampler:
+    """Base class for Megatron style BatchSampler."""
+
+    @abc.abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    @abc.abstractmethod
+    def __iter__(self):
+        ...
+
+    @property
+    @abc.abstractmethod
+    def num_micro_batch_times_micro_batch_size(self) -> int:
+        """The size of global batch on each data parallel rank."""
+        ...
+
+    @num_micro_batch_times_micro_batch_size.setter
+    @abc.abstractclassmethod
+    def num_micro_batch_times_micro_batch_size(self) -> None:
+        """The size of global batch on each data parallel rank."""
+        ...
+
+
+class MegatronPretrainingBatchSampler(BaseMegatronBatchSampler):
+    def __init__(
+        self,
+        total_samples: int,
+        consumed_samples: int,
+        num_micro_batch_times_micro_batch_size: int,
+        data_parallel_rank: int,
+        data_parallel_size: int,
+        drop_last: bool = True,
+    ) -> None:
+        # Sanity checks.
+        if total_samples <= 0:
+            raise RuntimeError('no sample to consume: {}'.format(self.total_samples))
+        if consumed_samples >= total_samples:
+            raise RuntimeError('no samples left to consume: {}, {}'.format(self.consumed_samples, self.total_samples))
+        if num_micro_batch_times_micro_batch_size <= 0:
+            raise RuntimeError(
+                f"num_micro_batch_times_micro_batch_size size must be greater than 0: {num_micro_batch_times_micro_batch_size}"
+            )
+        if data_parallel_size <= 0:
+            raise RuntimeError(f"data parallel size must be greater than 0: {data_parallel_size}")
+        if data_parallel_rank >= data_parallel_size:
+            raise RuntimeError(
+                'data_parallel_rank should be smaller than data size: {}, {}'.format(
+                    self.data_parallel_rank, data_parallel_size
+                )
+            )
+        # Keep a copy of input params for later use.
+        self.total_samples = total_samples
+        self.consumed_samples = consumed_samples
+        self._num_micro_batch_times_micro_batch_size = num_micro_batch_times_micro_batch_size
+        self.data_parallel_rank = data_parallel_rank
+        self.data_parallel_size = data_parallel_size
+        self.num_micro_batch_times_micro_batch_size_times_data_parallel_size = (
+            self._num_micro_batch_times_micro_batch_size * data_parallel_size
+        )
+        self.drop_last = drop_last
+
+    def __len__(self):
+        return self.total_samples
```

Comment:
Why is the `__len__` different from before?

Reply:
Earlier it was `return (self.total_samples - self.consumed_samples - 1) // self.micro_batch_times_data_parallel_size + 1`.

Reply:
nice catch. I think I just wrote this to meet the BatchSampler's requirements.